### PR TITLE
Add `wasi_ephemeral_nn` module support

### PR DESCRIPTION
**Pull request by AuYang261**
Add `wasi_ephemeral_nn` module support with optional cmake variable, which mentioned in #3229.

### CODE REVIEW

File: core/iwasm/common/wasm_native.c [hunk 1]
```diff
@@ -570,6 +570,12 @@ wasm_native_init()
     if (!wasm_native_register_natives("wasi_nn", native_symbols,
                                       n_native_symbols))
         goto fail;
+#if WASM_ENABLE_WASI_EPHEMERAL_NN != 0
+    n_native_symbols = get_wasi_ephemeral_nn_export_apis(&native_symbols);
+    if (!wasm_native_register_natives("wasi_ephemeral_nn", native_symbols,
+                                      n_native_symbols))
+        goto fail;
+#endif // WASM_ENABLE_WASI_EPHEMERAL_NN != 0
```
Comment (wenyongh) [inline core/iwasm/common/wasm_native.c:578]: Had better use `/* ... */` to comment the code in C source file.

File: core/iwasm/libraries/wasi-nn/src/utils/wasi_nn_app_native.c [hunk 2]
```diff
@@ -72,6 +72,51 @@ graph_builder_array_app_native(wasm_module_inst_t instance,
...
+    graph_builder *builder =
+        (graph_builder *)wasm_runtime_malloc(size * sizeof(graph_builder));
+    if (builder == NULL)
+        return missing_memory;
+
+    for (uint32_t i = 0; i < size; ++i) {
+        if (!wasm_runtime_validate_app_addr(
+                instance, builder_wasm[i].buf_offset, builder_wasm[i].size)) {
+            NN_ERR_PRINTF("builder_wasm[i].buf_offset is invalid");
+            return invalid_argument;
```
Comment (wenyongh) [inline core/iwasm/libraries/wasi-nn/src/utils/wasi_nn_app_native.c:99]: Should `wasm_runtime_free(builder)` before return

**Comment by wenyongh** [comment]
@tonibofarull could you help review this PR? Thanks a lot.

File: core/iwasm/libraries/wasi-nn/src/utils/wasi_nn_app_native.c (hunk 2 above)
Comment (AuYang261) [reply core/iwasm/libraries/wasi-nn/src/utils/wasi_nn_app_native.c:100]: Sorry, I forgot it.

**Review by wenyongh — COMMENTED** [review]
LGTM

File: core/iwasm/common/wasm_native.c [hunk 3]
```diff
@@ -570,6 +570,12 @@ wasm_native_init()
     if (!wasm_native_register_natives("wasi_nn", native_symbols,
                                       n_native_symbols))
         goto fail;
+#if WASM_ENABLE_WASI_EPHEMERAL_NN != 0
+    n_native_symbols = get_wasi_ephemeral_nn_export_apis(&native_symbols);
+    if (!wasm_native_register_natives("wasi_ephemeral_nn", native_symbols,
+                                      n_native_symbols))
+        goto fail;
+#endif /* WASM_ENABLE_WASI_EPHEMERAL_NN != 0 */
```
Comment (tonibofarull) [inline core/iwasm/common/wasm_native.c:578]: I strongly believe that we shouldn't replicate the code but instead define here `wasi_nn` or `wasi_ephemeral_nn` as part of the configuration.

**Review by tonibofarull — CHANGES_REQUESTED** [review]

File: core/iwasm/common/wasm_native.c (hunk 3 above)
Comment (lum1n0us) [reply core/iwasm/common/wasm_native.c:578]: Just FYI, in [wasi-nn](https://github.com/bytecodealliance/wasi-nn), both [rust api](https://github.com/bytecodealliance/wasi-nn/blob/9d825ab99921f92dfdee719e2b6d6aba366501dd/rust/src/generated.rs#L304) and assembly script api are using "wasi_ephemeral_nn" as its one and only import name.

Comment (tonibofarull) [reply core/iwasm/common/wasm_native.c:578]: Before deprecating `wasi_nn` I'd give the option to select one or the other.

Comment (AuYang261) [reply core/iwasm/common/wasm_native.c:578]: > Just FYI, in [wasi-nn](https://github.com/bytecodealliance/wasi-nn), both [rust api](https://github.com/bytecodealliance/wasi-nn/blob/9d825ab99921f92dfdee719e2b6d6aba366501dd/rust/src/generated.rs#L304) and assembly script api are using "wasi_ephemeral_nn" as its one and only import name.

Do you mean `wasi_nn` and `wasi_ephemeral_nn` only need to be registered one and only one?

Comment (lum1n0us) [reply core/iwasm/common/wasm_native.c:578]: > I strongly believe that we shouldn't replicate the code but instead define here `wasi_nn` or `wasi_ephemeral_nn` as part of the configuration.

It seems there are slight difference in this file between original `wasi_nn_xx()` and newly added `wasi_ephemeral_nn_xx()`. I am wondering can merge both?

Comment (tonibofarull) [reply core/iwasm/common/wasm_native.c:578]: I'd consider on supporting only 1 at a time and update the implementation of `wasi_nn_xx` and `wasi_ephemeral_nn`, if there are any.

**Comment by wenyongh** [comment]
@AuYang261 I have just merged PR #3246 to disable running CodeQL CI on each pull request, or it will run for a long time and fail eventually. Could you please rebase your branch with WAMR main?

**Review by tonibofarull — APPROVED** [review]
I'd like to verify later on why there's this drift from one to the other.

**Pull request merged**